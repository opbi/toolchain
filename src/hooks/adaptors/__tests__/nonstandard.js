import adaptorNonstandard from '../nonstandard';

import { eventLogger, errorRetry, chain } from '../../index';

const loggerMock = {
  info: jest.fn(),
  error: jest.fn(),
};

const [revertor, adaptor] = adaptorNonstandard;

describe('adaptorNonstandard', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('helps to use hook on non-standard functions with meta, context', async () => {
    const original = (a, b, c) => a + b + c;

    const adapted = chain(
      revertor(),
      eventLogger({ logParam: true }),
      adaptor(),
    )(original);

    const result = await adapted(1, 2, 3, {}, { logger: loggerMock });

    expect(result).toBe(6);
    expect(loggerMock.info.mock.calls).toMatchSnapshot();
  });

  it('helps to use hooks on non-standard functions without meta, context', async () => {
    const original = jest
      .fn((input) => input)
      .mockImplementationOnce(() => {
        throw Error('timeout');
      });

    const adapted = chain(revertor(), errorRetry(), adaptor())(original);

    const result = await adapted('yes');

    expect(result).toBe('yes');
    expect(original.mock.calls).toMatchSnapshot();
  });
});
