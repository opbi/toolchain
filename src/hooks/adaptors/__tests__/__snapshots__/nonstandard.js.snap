// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adaptorNonstandard helps to use hook on non-standard functions with meta, context 1`] = `
Array [
  Array [
    Object {
      "event": "original",
      "param": Array [
        1,
        2,
        3,
      ],
    },
    "original",
  ],
]
`;

exports[`adaptorNonstandard helps to use hooks on non-standard functions without meta, context 1`] = `
Array [
  Array [
    "yes",
    Object {},
    Object {},
  ],
  Array [
    "yes",
    Object {
      "maxRetries": 3,
      "retries": 1,
    },
    Object {},
  ],
]
`;
